### DISCUSSION TITLE
Slightly confused about UserData constructors

### DISCUSSION THREAD

**Discussion by Luminoth** [mlua-rs/mlua#170] [Q&A] [answered]
I'm looking at https://github.com/khvzak/mlua/blob/master/examples/userdata.rs and I see this in Rectangle::add_methods():

```
// Constructor
methods.add_meta_function(MetaMethod::Call, |_, ()| Ok(Rectangle::default()));
```

but in the usage of that it's doing this:

```
let rectangle = Rectangle::default();
lua.load(chunk! {
    local rect = $rectangle()
```

which seems wrong on the surface since I think it's creating 2 rectangles (one in Rust and then another in lua)?

Alternatively, looking at https://github.com/khvzak/mlua/blob/master/examples/guided_tour.rs I see this instead of defining a MetaMethod::Call constructor:

```
let vec2_constructor = lua.create_function(|_, (x, y): (f32, f32)| Ok(Vec2(x, y)))?;
globals.set("vec2", vec2_constructor)?;
```

and that makes sense but now I don't know what the right way to do this is between the two. If I just do this:

```
impl UserData for Animal {
    fn add_methods<'lua, M: UserDataMethods<'lua, Self>>(methods: &mut M) {
        // constructor
        methods.add_meta_function(
            MetaMethod::Call,
            |_, (noise_emitted, num_legs): (String, usize)| {
                Ok(Animal::new(noise_emitted, num_legs))
            },
        );
        methods.add_method("speak", |_, this, ()| Ok(this.speak()));
        methods.add_method("num_legs", |_, this, ()| Ok(this.num_legs()));
    }
}

fn userdata_example() -> anyhow::Result<()> {
    let lua = Lua::new();

    lua.load(chunk! {
        local cat = Animal("Meow", 4);
        print("[Lua]: A cat has " .. cat:num_legs() .. " legs");
    })
    .exec()?;

    Ok(())
}
```

I get this error: ```attempt to call a nil value (global 'Animal')``` so I'm definitely doing something wrong here, but I'm not sure what the actual correct way to resolve this is.

**Comment by khvzak** [comment, marked as answer]
> which seems wrong on the surface since I think it's creating 2 rectangles (one in Rust and then another in lua)?

It's true but you need an instance of Userdata to exec `__call` OR just a new function to construct it (as it's done in `vec2_constructor `).

In your example `Animal` is just a Rust type that implements `UserData` but in Lua it **must** be an entity in a global namespace to use it on that way.
It your case it should be a "dummy" instance of `Animal` or you can make it as a function and remove `__call` constructor.  
In another words - an existing instance of Animal that constructs new instances of self.

**Reply by Luminoth** [reply]
if I make it a function, is there any way to register that from the UserData itself? I'm not seeing any way just looking at the trait and it seems odd to expect the user of my UserData to always have to add it manually.
